packages/progress-monitor/src/browser/progress-dialog.ts

```diff
@@ -0,0 +1,92 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from 'inversify';
+import { Message } from '@phosphor/messaging';
+import { Disposable } from '@theia/core';
+import { Widget, BaseWidget, Key } from '@theia/core/lib/browser';
+import { ProgressWidget } from './progress-widget';
+
+@injectable()
+export class ProgressDialog extends BaseWidget {
+
+    protected contentNode: HTMLDivElement;
+
+    protected isOpen: boolean = false;
+
+    @inject(ProgressWidget) readonly widget: ProgressWidget;
+
+    constructor() {
+        super();
+        this.contentNode = document.createElement('div');
+        this.contentNode.classList.add(ProgressWidget.Styles.PROGRESS_MONITOR_CONTENT);
+
+        this.node.classList.add(ProgressWidget.Styles.PROGRESS_WIDGET_DIALOG);
+
+        this.node.appendChild(this.contentNode);
+        this.update();
+    }
+
+    @postConstruct()
+    protected init() {
+        this.toDispose.push(this.widget);
+    }
+
+    protected onUpdateRequest(msg: Message): void {
+        super.onUpdateRequest(msg);
+        this.widget.update();
+    }
+
+    protected onAfterAttach(msg: Message): void {
+        super.onAfterAttach(msg);
+        Widget.attach(this.widget, this.contentNode);
+        this.toDisposeOnDetach.push(Disposable.create(() =>
+            Widget.detach(this.widget)
+        ));
+
+        this.addKeyListener(document.body, Key.ESCAPE, e => {
+            this.closeDialog();
+        });
+        this.addEventListener(document.body, 'click', e => {
+            const thisStatusBarItem = document.getElementsByClassName(ProgressWidget.Styles.PROGRESS_STATUS_BAR).item(0) as HTMLElement;
+            if (e.srcElement && !thisStatusBarItem.contains(e.srcElement)) {
+                this.closeDialog();
+            }
+        });
+    }
+
+    closeDialog() {
+        this.isOpen = false;
+        this.close();
+    }
+
+    protected onActivateRequest(msg: Message): void {
+        super.onActivateRequest(msg);
+        this.widget.activate();
+    }
+
+    async toggleOpen(): Promise<void> {
+        if (this.isOpen) {
+            this.closeDialog();
+        } else {
+            const thisStatusBarItem = document.getElementsByClassName(ProgressWidget.Styles.PROGRESS_STATUS_BAR).item(0) as HTMLElement;
```

Comment:
What is happening here? It is strange to make assumptions about the implementation of the status bar. If one changes how it is rendered this code is going to be broken.

Reply:
It is making the progress dialog and the status bar be under the same div so that when the items on the status bar moves the dialog always stays above the status bar item. 

> It is strange to make assumptions about the implementation of the status bar. If one changes how it is rendered this code is going to be broken.

I'm okay with changing this I just think its going to be hard to keep the dialog over the status bar item. I don't think we should set some absolute position for the progress dialog in the bottom right corner for example, it would be weird to click the status bar item in one place and have the dialog appear in another.

Reply:
ok, I thought `MouseEvent` contains coordinates and it should be possible to display a node at them.

Reply:
I think that can work but then you would have to control what happens when something on the status bar finishes and the status bar item moves. I think in that case the dialog would stay in the old location unless you somehow got the location of the status bar element (which brings us back to the first issue).

Reply:
I see. My main concern is breaking the encapsulation of the status bar. @svenefftinge We could provide add `statusBar.getNode(id): HTMLElement` to hide how status bar entry is rendered.

Reply:
> statusBar.getNode(id): HTMLElement | undefined

I think that would be a great way to do it!